### ISSUE TITLE
Swarm plugin-related error "plugin does not implement PluginAddr interface" fills disk

### ISSUE THREAD

**Issue by jonaslb** [moby/moby#48133]
### Description

With docker 26.1 we saw the following logs being repeated so often that disk space literally filled up (syslog):

```
2024-05-08T10:49:19.381166+00:00 hostnamereplaced dockerd[25865]: time="2024-05-08T10:49:19.381077536Z" level=info msg="attempting to unpublish volume" attempt=436364316 module=node/agent/csi volume.id=jjn8qz56mg3852jdllefpz21l
2024-05-08T10:49:19.381171+00:00 hostnamereplaced dockerd[25865]: time="2024-05-08T10:49:19.381115531Z" level=info msg="upublishing volume failed" attempt=436364316 error="cannot get plugin csi-smb:latest: plugin does not implement PluginAddr interface" module=node/agent/csi volume.id=jjn8qz56mg3852jdllefpz21l 
```

I think this is possibly related to moby/swarmkit#3171, which seems to have altered the logic that decides to log this error.

The cluster volume driver we use is the kubernetes SMB CSI driver, built for swarm according to information from <https://github.com/olljanat/csi-plugins-for-docker-swarm>.

The nodes that we would see this error message on were not supposed to have any containers running with the volume mounted, so we don't actually understand why it is attempting to unpublish it. 

### Reproduce

We can not trigger reproduction actively, it just "seems to happen" after some time. See description above.

### Expected behavior

First of all it seems that even if some error occurs, it should not proceed to busy-loop and log every iteration.

Second, the log seems to indicate a programming error - some incompatibility between swarmkit and docker itself?

### docker version

```bash
It was docker 26.1. We have since downgraded to 26.0 / some nodes updated to 27, which are not relevant to this issue.
```

### docker info

```bash
See docker version info: We have since downgraded and i did not get the info while we were on the bad version.
```

**Comment by thaJeztah** [comment]
cc @dperny PTAL

**Comment by beornf** [comment]
I can confirm that I saw this swarm bug with a CSI volume plugin and forgot to report it. I am currently using Docker v26.0.2 which doesn't have this regression.

**Comment by olljanat** [comment]
@blib Are you sure that https://github.com/moby/swarmkit/pull/3181 fixes this issue? As far I see issue still exists in version 27.4.0-rc.2 even when https://github.com/moby/moby/pull/48687 included your change to it.

Latest known good version of CSI users still is v26.0.2

**Comment by thaJeztah** [comment]
I wonder if the issue is somewhere in this area. At least this seems to be code that's called to get the plugin; https://github.com/moby/moby/blob/d3cc051f50733848f153ccf6a99259acec90d265/daemon/cluster/convert/pluginadapter.go#L27-L33

Which returns a `swarmPlugin`;
https://github.com/moby/moby/blob/d3cc051f50733848f153ccf6a99259acec90d265/daemon/cluster/convert/pluginadapter.go#L19-L25

Which looks to implement `PluginWithV1Client`, and wraps `CompatPlugin`, but none of those seem to implement the `PluginAddr` interface?
https://github.com/moby/moby/blob/d3cc051f50733848f153ccf6a99259acec90d265/pkg/plugingetter/getter.go#L19-L30

**Comment by florian-forestier** [comment]
Hi,

Looking in the code, it sounds like this commit https://github.com/moby/swarmkit/commit/efe042779e7d87d9765cbd6a6799449197cd5cc2 may be interesting to understand what happened between 26.0.2 and 26.1.0. 

Between 26.0.2 and 26.1.0, `moby/swarmkit` dependency was updated from `v2.0.0-20240125134710-dcda100a8261` to `v2.0.0-20240412154004-f3ffc0881d0e`; and commit [moby/swarmkit#efe0427](https://github.com/moby/swarmkit/commit/efe042779e7d87d9765cbd6a6799449197cd5cc2) was added in the meantime. It seems that the plugin registration method changed, removing the `plugingetter.CompatPlugin`.

Unfortunately I don't currently have much time to spend on it, but will look again on it later on the week.

**Comment by olljanat** [comment]
> I think there's a missing compatbility layer between plugingetter.PluginAddr and plugin.AddrPlugin

Sounds reasonable explanation to me.

Ping @corhere in case you still remember https://github.com/moby/swarmkit/commit/efe042779e7d87d9765cbd6a6799449197cd5cc2 ?


PS. Also it would make sense to reopen this issue by someone how have rights to do so.

**Comment by florian-forestier** [comment]
Do we consider this is a bug in moby/moby directly (meaning that moby have to support cast from "swarm" plugin's own implementation), or is it a bug in moby/swarmkit that should be fixed directly in swarm?

I think it makes more sense to fix in swarmkit rather than in moby. WDYT?

**Comment by florian-forestier** [comment]
I presume that Enterprise Swarm does not have this problem... 😏

**Comment by blib** [comment]
there is no reason to stay with docker ... try OrbStack or k3s.

**Comment by olljanat** [comment]
> I presume that Enterprise Swarm does not have this problem... 😏

Yes, latest Mirantis Container Runtime seems to be [25.0.10](https://docs.mirantis.com/mcr/25.0/release-notes.html) which should same what you get when building binaries from moby branch [25.0](https://github.com/moby/moby/commits/25.0)

Additionally, afaiu CSI support is still considered as experimental and it starts to look for me than they don’t have plans to finalize it so most likely Enterprise customers are not using those.

So at the moment my recommendation is to use non-CSI volume plugins for any critical workloads.

**Comment by olljanat** [comment]
> I think there's a missing compatbility layer between `plugingetter.PluginAddr` and `plugin.AddrPlugin`

@florian-forestier As far I was able to track, it is https://github.com/moby/moby/blob/master/daemon/cluster/convert/pluginadapter.go which should act as compatibility layer but clearly something is missing from it.
